File: frontend/lib/ipc/dto/enums.dart
// Dart mirror enums for the backend data-model wire contract.
//
// Every value below MUST equal the backend `#[serde(rename_all = "snake_case")]` wire value
// (Stučka/backend/crates/data-model/src/enums.rs + pii.rs). The round-trip is asserted in
// test/dto/enum_roundtrip_test.dart against fixtures lifted verbatim from the Rust tests.

/// A serde-snake_case enum mirror with explicit wire values.
mixin _WireEnum {
  String get wire;
}

/// IdentityType — 11 forced single-choice (data-model enums.rs::IdentityType).
enum IdentityType implements _WireEnum {
  standardFullTime('standard_full_time'),
  dispatch('dispatch'),
  partTime('part_time'),
  newEmployment('new_employment'),
  domesticService('domestic_service'),
  constructionLabor('construction_labor'),
  individualEmployee('individual_employee'),
  intern('intern'),
  retiredRehired('retired_rehired'),
  contractor('contractor'),
  deFactoNoContract('de_facto_no_contract');

  const IdentityType(this.wire);
  @override
  final String wire;

  String get labelZh => switch (this) {
        standardFullTime => '城镇标准全日制',
        dispatch => '劳务派遣',
        partTime => '非全日制',
        newEmployment => '网约车 / 外卖 / 即时配送',
        domesticService => '家政 / 月嫂 / 钟点工',
        constructionLabor => '建筑劳务（农民工）',
        individualEmployee => '个体工商户雇员',
        intern => '实习生 / 学徒',
        retiredRehired => '退休返聘',
        contractor => '承包 / 承揽 / 自由职业',
        deFactoNoContract => '未签合同事实劳动',
      };

  static IdentityType fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => standardFullTime);
}

/// DisputeSubtype — 8 social-insurance branches (data-model enums.rs::DisputeSubtype).
///
/// NOTE (brief §6.4): the backend currently only models the 8 social-insurance subtypes; the
/// 农民工欠薪 (WAGE_ARREARS) main path is not yet a backend enum value. `IntakePage` renders the
/// real backend enum; WAGE_ARREARS stays a placeholder until data-model extends the enum.
enum DisputeSubtype implements _WireEnum {
  socialInsWaiverInvalid('social_ins_waiver_invalid'),
  socialInsArrears('social_ins_arrears'),
  socialInsUnderpaidBase('social_ins_underpaid_base'),
  socialInsIntermittent('social_ins_intermittent'),
  socialInsProxy('social_ins_proxy'),
  socialInsUninsuredInjury('social_ins_uninsured_injury'),
  socialInsCrossPeriod('social_ins_cross_period'),
  socialInsBaseDispute('social_ins_base_dispute');

  const DisputeSubtype(this.wire);
  @override
  final String wire;

  String get labelZh => switch (this) {
        socialInsWaiverInvalid => '弃缴约定无效 + 单方解除',
        socialInsArrears => '社保普通欠缴',
        socialInsUnderpaidBase => '低缴',
        socialInsIntermittent => '漏缴',
        socialInsProxy => '挂靠社保',
        socialInsUninsuredInjury => '未参保导致工伤损失',
        socialInsCrossPeriod => '跨期争议',
        socialInsBaseDispute => '基数缴费争议',
      };

  static DisputeSubtype fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => socialInsArrears);
}

enum CoverageTier implements _WireEnum {
  makeDeep('make_deep'),
  makeUsable('make_usable');

  const CoverageTier(this.wire);
  @override
  final String wire;

  static CoverageTier fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => makeUsable);
}

enum CaseStatus implements _WireEnum {
  draft('draft'),
  diagnosed('diagnosed'),
  confirmed('confirmed'),
  frozen('frozen'),
  disputed('disputed');

  const CaseStatus(this.wire);
  @override
  final String wire;

  static CaseStatus fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => draft);
}

enum FactCategory implements _WireEnum {
  relationQualification('relation_qualification'),
  wage('wage'),
  timePeriod('time_period'),
  terminationReason('termination_reason'),
  workInjury('work_injury'),
  discrimination('discrimination'),
  other('other');

  const FactCategory(this.wire);
  @override
  final String wire;

  static FactCategory fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => other);
}

enum FactStatus implements _WireEnum {
  pending('pending'),
  confirmed('confirmed'),
  disputed('disputed'),
  deprecated('deprecated');

  const FactStatus(this.wire);
  @override
  final String wire;

  static FactStatus fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => pending);
}

enum FactSource implements _WireEnum {
  userInput('user_input'),
  aiInferred('ai_inferred'),
  ruleEngine('rule_engine');

  const FactSource(this.wire);
  @override
  final String wire;

  static FactSource fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => userInput);
}

/// EvidenceCategory — 7 classes (data-model enums.rs::EvidenceCategory).
enum EvidenceCategory implements _WireEnum {
  documentaryContract('documentary_contract'),
  audioVideo('audio_video'),
  digitalCommunication('digital_communication'),
  witnessStatement('witness_statement'),
  scenePhotoVideo('scene_photo_video'),
  thirdPartyData('third_party_data'),
  appraisal('appraisal');

  const EvidenceCategory(this.wire);
  @override
  final String wire;

  String get labelZh => switch (this) {
        documentaryContract => '书证 / 合同',
        audioVideo => '录音 / 录像',
        digitalCommunication => '电子聊天 / 通讯记录',
        witnessStatement => '证人证言',
        scenePhotoVideo => '现场照片 / 视频',
        thirdPartyData => '银行流水 / 税单 / 社保',
        appraisal => '鉴定 / 评估',
      };

  static EvidenceCategory fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => documentaryContract);
}

enum EvidenceStatus implements _WireEnum {
  uploaded('uploaded'),
  parsed('parsed'),
  scored('scored'),
  verified('verified'),
  disputed('disputed'),
  quarantined('quarantined');

  const EvidenceStatus(this.wire);
  @override
  final String wire;

  static EvidenceStatus fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => uploaded);
}

enum ClaimType implements _WireEnum {
  economicCompensation('economic_compensation'),
  economicDamage('economic_damage'),
  doubleWageNoContract('double_wage_no_contract'),
  overtimePay('overtime_pay'),
  maliciousArrearsSurcharge('malicious_arrears_surcharge'),
  workInjuryBenefit('work_injury_benefit'),
  other('other');

  const ClaimType(this.wire);
  @override
  final String wire;

  static ClaimType fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => other);
}

enum ClaimStatus implements _WireEnum {
  draft('draft'),
  finalized('finalized'),
  granted('granted'),
  denied('denied'),
  withdrawn('withdrawn');

  const ClaimStatus(this.wire);
  @override
  final String wire;

  static ClaimStatus fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => draft);
}

enum CoverageTag implements _WireEnum {
  exact('exact'),
  approximate('approximate'),
  boundary('boundary'),
  unknown('unknown');

  const CoverageTag(this.wire);
  @override
  final String wire;

  static CoverageTag fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => unknown);
}

/// SourceTag — answer provenance (data-model enums.rs::SourceTag).
/// Wire values are `online` (NOT `web`) and `inferred` (NOT `infer`).
enum SourceTag implements _WireEnum {
  rule('rule'),
  kb('kb'),
  online('online'),
  inferred('inferred');

  const SourceTag(this.wire);
  @override
  final String wire;

  String get labelZh => switch (this) {
        rule => '规则',
        kb => '知识库',
        online => '联网',
        inferred => '推断',
      };

  static SourceTag fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => inferred);
}

enum AuditCategory implements _WireEnum {
  aiSuggestion('ai_suggestion'),
  userReview('user_review'),
  stateChange('state_change'),
  mergeDecision('merge_decision'),
  cryptoOp('crypto_op'),
  export('export'),
  sync('sync');

  const AuditCategory(this.wire);
  @override
  final String wire;

  static AuditCategory fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => stateChange);
}

enum LawLevel implements _WireEnum {
  constitution('constitution'),
  statute('statute'),
  judicialInterpretation('judicial_interpretation'),
  administrativeRegulation('administrative_regulation'),
  departmentalRule('departmental_rule'),
  localRegulation('local_regulation');

  const LawLevel(this.wire);
  @override
  final String wire;

  static LawLevel fromWire(String w) =>
      values.firstWhere((e) => e.wire == w, orElse: () => statute);
}

File: backend/crates/data-model/src/enums.rs
//! First-class object enumerations (data/01 §1.0.4 / §1.2 / §1.3 / §1.5, ai/02 §2, ai/03 §3).
//!
//! Every enum serialises `rename_all = "snake_case"` for the D1 Bearer-HTTP JSON contract.
//! `sqlx::Type` derives are gated behind the optional `sqlx` feature via `cfg_attr` so the
//! db layer can persist them while rule-engine never depends on sqlx (INV-01 boundary).
//!
//! The `sqlx(type_name = ...)` attribute names the Postgres ENUM type; under the SQLite
//! fallback the value is stored as `TEXT` and the rename_all snake_case form is the
//! on-disk representation.

use serde::{Deserialize, Serialize};

/// Identity type — 11 forced single-choice (data/01 §1.0.4, prd §3.1.1). R1 immutable.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "identity_type", rename_all = "snake_case")
)]
pub enum IdentityType {
    StandardFullTime,   // 01 城镇标准全日制
    Dispatch,           // 02 劳务派遣
    PartTime,           // 03 非全日制
    NewEmployment,      // 04 网约车 / 外卖 / 即时配送
    DomesticService,    // 05 家政 / 月嫂 / 钟点工
    ConstructionLabor,  // 06 建筑劳务（农民工）
    IndividualEmployee, // 07 个体工商户雇员
    Intern,             // 08 实习生 / 学徒
    RetiredRehired,     // 09 退休返聘
    Contractor,         // 10 承包 / 承揽 / 自由职业
    DeFactoNoContract,  // 11 未签合同事实劳动
}

/// Dispute subtype — 8 social-insurance branches (data/01 §1.1.4, prd §3.3.2).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "dispute_subtype", rename_all = "snake_case")
)]
pub enum DisputeSubtype {
    SocialInsWaiverInvalid,   // 1 弃缴约定无效 + 单方解除（法释 12 号 §19）
    SocialInsArrears,         // 2 普通欠缴
    SocialInsUnderpaidBase,   // 3 低缴（abstention）
    SocialInsIntermittent,    // 4 漏缴
    SocialInsProxy,           // 5 挂靠社保（abstention）
    SocialInsUninsuredInjury, // 6 未参保导致工伤损失
    SocialInsCrossPeriod,     // 7 跨期争议（abstention）
    SocialInsBaseDispute,     // 8 基数缴费争议（abstention）
}

/// Coverage tier — 做深 / 做能用 (data/01 §1.1.4).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "coverage_tier", rename_all = "snake_case")
)]
pub enum CoverageTier {
    MakeDeep, // 做深
    #[default]
    MakeUsable, // 做能用（default）
}

/// Case lifecycle status (data/01 §1.1.3 state machine).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "case_status", rename_all = "snake_case")
)]
pub enum CaseStatus {
    #[default]
    Draft,
    Diagnosed,
    Confirmed,
    Frozen,
    Disputed,
}

/// Fact category (data/01 §1.2.4).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "fact_category", rename_all = "snake_case")
)]
pub enum FactCategory {
    RelationQualification, // 关系定性
    Wage,                  // 工资
    TimePeriod,            // 时间
    TerminationReason,     // 解除原因
    WorkInjury,            // 工伤
    Discrimination,        // 歧视
    Other,                 // 其他
}

/// Fact lifecycle status (data/01 §1.2.3 state machine).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "fact_status", rename_all = "snake_case")
)]
pub enum FactStatus {
    #[default]
    Pending,
    Confirmed,
    Disputed,
    Deprecated,
}

/// Fact provenance (data/01 §1.2.4).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "fact_source", rename_all = "snake_case")
)]
pub enum FactSource {
    UserInput,
    AiInferred,
    RuleEngine,
}

/// Evidence category — 7 classes (data/01 §1.3.3).
///
/// Named `EvidenceCategory` per the API contract (backend/01 `EvidenceCategory`); the
/// data/01 Postgres ENUM type name remains `evidence_type`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "evidence_type", rename_all = "snake_case")
)]
pub enum EvidenceCategory {
    DocumentaryContract,  // 1 书证 / 合同
    AudioVideo,           // 2 录音 / 录像
    DigitalCommunication, // 3 电子聊天 / 通讯记录
    WitnessStatement,     // 4 证人证言
    ScenePhotoVideo,      // 5 现场照片 / 视频
    ThirdPartyData,       // 6 银行流水 / 税单 / 社保
    Appraisal,            // 7 鉴定 / 评估（工伤 / 职业病 / 伤残）
}

/// Evidence lifecycle status (data/01 §1.3.4 state machine).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "evidence_status", rename_all = "snake_case")
)]
pub enum EvidenceStatus {
    #[default]
    Uploaded,
    Parsed,
    Scored,
    Verified,
    Disputed,
    Quarantined,
}

/// Claim type (data/01 §1.5.4).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "claim_type", rename_all = "snake_case")
)]
pub enum ClaimType {
    EconomicCompensation,      // 经济补偿
    EconomicDamage,            // 经济赔偿
    DoubleWageNoContract,      // 二倍工资
    OvertimePay,               // 加班费
    MaliciousArrearsSurcharge, // 50% 加付
    WorkInjuryBenefit,         // 工伤待遇
    Other,
}

/// Claim lifecycle status (data/01 §1.5.3 state machine).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "claim_status", rename_all = "snake_case")
)]
pub enum ClaimStatus {
    #[default]
    Draft,
    Finalized,
    Granted,
    Denied,
    Withdrawn,
}

/// Coverage tag — rule-engine source-coverage classification (ai/02 §2, INV-01).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "coverage_tag", rename_all = "snake_case")
)]
pub enum CoverageTag {
    Exact,       // 法源精确匹配
    Approximate, // 法源大致适用
    Boundary,    // 法源边界条件
    Unknown,     // 无法源支撑，必须 abstention
}

/// Source tag — answer provenance for INV-08 disclosure (ai/03 §3.1, `crates/data-model/src/source_tag.rs`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "source_tag", rename_all = "snake_case")
)]
pub enum SourceTag {
    Rule,     // 规则引擎（M9 / M5）—— INV-01 优先
    Kb,       // 本地知识库（带 kb_version 冻结 hash）
    Online,   // 联网检索（带 provider + ts）
    Inferred, // 模型推断（最低可信度）
}

/// Audit category projection (backend/01 §1.8 `AuditCategory`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "audit_category", rename_all = "snake_case")
)]
pub enum AuditCategory {
    AiSuggestion,
    UserReview,
    StateChange,
    MergeDecision,
    CryptoOp,
    Export,
    Sync,
}

/// Next-action kind suggested when the rule engine / deadline engine abstains (ai/03 §3.4,
/// shared by rule-engine + ai-dispatcher). The Low-bucket key-facts are populated from these.
///
/// data/01 carries no associated payload here so the value can be a `sqlx::Type` ENUM; the
/// per-variant detail (field name / evidence class / hint text) is carried alongside in the
/// dispatcher's `next_actions` payload, keeping this shared enum a plain discriminant.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "next_action", rename_all = "snake_case")
)]
pub enum NextAction {
    CollectFact,     // 缺事实
    CollectEvidence, // 缺证据
    ShowUiHint,      // UI 提示（如 X 阈值免责）
    ManualConfirm,   // M5 时效 buffer 强制人工二次确认（INV-08 锚）
    ConsultLegalAid, // 引导法援热线
}

/// Law hierarchy level (backend/01 §1.7 `LawLevel`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(
    feature = "sqlx",
    sqlx(type_name = "law_level", rename_all = "snake_case")
)]
pub enum LawLevel {
    Constitution,             // 宪法
    Statute,                  // 法律
    JudicialInterpretation,   // 司法解释
    AdministrativeRegulation, // 行政法规
    DepartmentalRule,         // 部门规章
    LocalRegulation,          // 地方性法规
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn identity_type_serializes_snake_case() {
        let s = serde_json::to_string(&IdentityType::DeFactoNoContract).unwrap();
        assert_eq!(s, "\"de_facto_no_contract\"");
        let s = serde_json::to_string(&IdentityType::StandardFullTime).unwrap();
        assert_eq!(s, "\"standard_full_time\"");
    }

    #[test]
    fn dispute_subtype_serializes_snake_case() {
        let s = serde_json::to_string(&DisputeSubtype::SocialInsWaiverInvalid).unwrap();
        assert_eq!(s, "\"social_ins_waiver_invalid\"");
    }

    #[test]
    fn case_status_roundtrip() {
        for st in [
            CaseStatus::Draft,
            CaseStatus::Diagnosed,
            CaseStatus::Confirmed,
            CaseStatus::Frozen,
            CaseStatus::Disputed,
        ] {
            let s = serde_json::to_string(&st).unwrap();
            let back: CaseStatus = serde_json::from_str(&s).unwrap();
            assert_eq!(st, back);
        }
        assert_eq!(
            serde_json::to_string(&CaseStatus::Frozen).unwrap(),
            "\"frozen\""
        );
    }

    #[test]
    fn defaults_match_spec() {
        assert_eq!(CoverageTier::default(), CoverageTier::MakeUsable);
        assert_eq!(CaseStatus::default(), CaseStatus::Draft);
        assert_eq!(FactStatus::default(), FactStatus::Pending);
        assert_eq!(EvidenceStatus::default(), EvidenceStatus::Uploaded);
        assert_eq!(ClaimStatus::default(), ClaimStatus::Draft);
    }

    #[test]
    fn evidence_category_seven_classes_snake_case() {
        assert_eq!(
            serde_json::to_string(&EvidenceCategory::DocumentaryContract).unwrap(),
            "\"documentary_contract\""
        );
        assert_eq!(
            serde_json::to_string(&EvidenceCategory::ThirdPartyData).unwrap(),
            "\"third_party_data\""
        );
    }

    #[test]
    fn coverage_and_source_tags_snake_case() {
        assert_eq!(
            serde_json::to_string(&CoverageTag::Unknown).unwrap(),
            "\"unknown\""
        );
        assert_eq!(
            serde_json::to_string(&SourceTag::Online).unwrap(),
            "\"online\""
        );
    }

    #[test]
    fn claim_type_snake_case() {
        assert_eq!(
            serde_json::to_string(&ClaimType::MaliciousArrearsSurcharge).unwrap(),
            "\"malicious_arrears_surcharge\""
        );
        assert_eq!(
            serde_json::to_string(&ClaimType::DoubleWageNoContract).unwrap(),
            "\"double_wage_no_contract\""
        );
    }

    #[test]
    fn audit_category_and_law_level_snake_case() {
        assert_eq!(
            serde_json::to_string(&AuditCategory::MergeDecision).unwrap(),
            "\"merge_decision\""
        );
        assert_eq!(
            serde_json::to_string(&LawLevel::JudicialInterpretation).unwrap(),
            "\"judicial_interpretation\""
        );
    }

    #[test]
    fn next_action_snake_case_roundtrip() {
        assert_eq!(
            serde_json::to_string(&NextAction::CollectFact).unwrap(),
            "\"collect_fact\""
        );
        assert_eq!(
            serde_json::to_string(&NextAction::ConsultLegalAid).unwrap(),
            "\"consult_legal_aid\""
        );
        for a in [
            NextAction::CollectFact,
            NextAction::CollectEvidence,
            NextAction::ShowUiHint,
            NextAction::ManualConfirm,
            NextAction::ConsultLegalAid,
        ] {
            let s = serde_json::to_string(&a).unwrap();
            let back: NextAction = serde_json::from_str(&s).unwrap();
            assert_eq!(a, back);
        }
    }
}
